lib/rules/no-unused-class-component-methods.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * @fileoverview Prevent declaring unused methods of component class
+ * @author Paweł Nowak
+ */
+
+'use strict';
+
+const Components = require('../util/Components');
+const astUtil = require('../util/ast');
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+const internalMethods = [
+  'constructor',
+  'componentDidCatch',
+  'componentDidMount',
+  'componentDidUpdate',
+  'componentWillMount',
+  'componentWillReceiveProps',
+  'componentWillUnmount',
+  'componentWillUpdate',
+  'getSnapshotBeforeUpdate',
+  'render',
+  'shouldComponentUpdate',
+  'UNSAFE_componentWillMount',
+  'UNSAFE_componentWillReceiveProps',
+  'UNSAFE_componentWillUpdate'
+];
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent declaring unused methods of component class',
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('no-unused-class-component-methods')
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false
+      }
+    ]
+  },
+
+  create: Components.detect((context, components, utils) => {
+    const isNotComponent = node => (
+      !utils.isES5Component(node) &&
+      !utils.isES6Component(node) &&
+      !utils.isCreateElement(node)
+    );
+    const filterAllMethods = node => {
+      const isMethod = node.type === 'MethodDefinition';
+      const isArrowFunction = (
+        node.type === 'ClassProperty' &&
+        node.value.type === 'ArrowFunctionExpression'
```

Comment:
what about async functions, async arrow functions, generators, and async generators?